---
title: "statistical_genetics_1"
output: html_document
---

## library
```{r}
library(dplyr)

```

## loading the data
```{r}
fms <- read.delim(file="/Users/jerome/Documents/ebook/FMS_data.txt", header=T, sep="\t")\
head(fms)
```

## Data manipulation
```{r}
attach(fms)
```


## Summary
### identify the minor allel and its frequency
```{r}
# The table() function in R outputs the counts of each level of the ordinal
# variable given as its argument.
GenoCount <- summary(actn3_rs540874)
GenoCount

# total number
NumbObs <- sum(!is.na(actn3_rs540874))

# geno freq
GenoFreq <- as.vector(GenoCount/NumbObs)
GenoFreq
```

```{r}
FreqA <- (2*GenoFreq[1] + GenoFreq[2])/2
FreqA
```

```{r}
FreqG <- (GenoFreq[2] + 2*GenoFreq[3])/2
FreqG
```

```{r}
# Thus, we report A is the minor allele at this SNP locus, with a frequency
# of 0.43.
```

```{r}
#install.packages("genetics")
library(genetics)
Geno <- genotype(actn3_rs540874,sep="")
summary(Geno)
```

## test for association
```{r}
NamesEsr1Snps <- names(fms)[substr(names(fms),1,4)=="esr1"]
NamesEsr1Snps

# The genotype matrix can now be defined by selecting the columns of fms that
# correspond to the esr1 SNP names
fmsEsr1 <- fms[,is.element(names(fms),NamesEsr1Snps)]
```

## trait
```{r}
# We define our trait to be an indicator for whether BMI is > 25 at baseline:
Trait <- as.numeric(pre.BMI>25)

```

## contigency table
```{r}
ObsTab <- table(Trait,Geno)
```

## chisq
```{r}
# test function
chisq_test <- function(Geno) {
    ObsTab <- table(Trait,Geno)
    return(chisq.test(ObsTab)$p.value)
}

table(fmsEsr1$esr1_rs1801132)
apply(fmsEsr1,2,chisq_test)
```

## fisher exact test
```{r}
# fisher test
fisher_test <- function(Geno){
    ObsTab <- table(Trait,Geno)
    return(fisher.test(ObsTab)$p.value)
}

apply(fmsEsr1,2,fisher_test)
```


## quantitative trait
```{r}
NamesResistinSnps <- names(fms)[substr(names(fms),1,8)=="resistin"]
fmsResistin <- fms[,is.element(names(fms),NamesResistinSnps)]
```
create a new function that takes a single genotype vector, converts
it to binary elements, and generates p-values based on the t-test of equality
of the mean trait across the resulting two levels. We first call the genetics
package since we are using the allele.names() function within this package

```{r}
library(genetics)
TtestPval <- function(Geno){
    alleleMajor <- allele.names(genotype(Geno, sep="",
    reorder="freq"))[1]
    GenoWt <- paste(alleleMajor, alleleMajor, sep="")
    GenoBin <- as.numeric(Geno!=GenoWt)[!is.na(Geno)]
    Trait <- NDRM.CH[!is.na(Geno)]
    return(t.test(Trait[GenoBin==1], Trait[GenoBin==0])$p.value)
}

# genoBIn binary vector indicating whether the geno type is wild type or not
# run the test
apply(fmsResistin,2,TtestPval)

```

## take a closer look at the association test
```{r}
Geno <- fms$"resistin_c180g"
table(Geno)
GenoWt <- names(table(Geno))[table(Geno)==max(table(Geno))]
GenoBin <- as.numeric(Geno!=GenoWt)[!is.na(Geno)]
Trait <- NDRM.CH[!is.na(Geno)]

# run the specific test 
t.test(Trait[GenoBin==1],Trait[GenoBin==0])

```


# replace the t-test with wilcoxon rank-sum test
```{r}
RankSumtestPval <- function(Geno){
    alleleMajor <- allele.names(genotype(Geno, sep="",
    reorder="freq"))[1]
    GenoWt <- paste(alleleMajor, alleleMajor, sep="")
    GenoBin <- as.numeric(Geno!=GenoWt)[!is.na(Geno)]
    Trait <- NDRM.CH[!is.na(Geno)]
    return(wilcox.test(Trait[GenoBin==1], Trait[GenoBin==0])$p.value)
}

# genoBIn binary vector indicating whether the geno type is wild type or not
# run the test
apply(fmsResistin,2,RankSumtestPval)
```


If a priori dichotomization of the genotype variables is not desirable, we
can perform an analysis of variance (ANOVA) or the non-parametric analog,
the Kruskal-Wallis test, to characterize association with a quantitative
trait.

# ANOVA 
The F-test in one-way analysis of variance is used to assess whether the expected values of a quantitative variable within several pre-defined groups differ from each other. For example, suppose that a medical trial compares four treatments. The ANOVA F-test can be used to assess whether any of the treatments is on average superior, or inferior, to the others versus the null hypothesis that all four treatments yield the same mean response.

```{r}
Geno <- as.factor(resistin_c180g)
Trait <- NDRM.CH

```

We use the lm() function in R to perform ANOVA. Alternatively, the
aov() function can be applied, though the corresponding output of the associated
print() and summary() functions will differ. Here we indicate that
we want to exclude individuals with missing values for the trait, coded as NA,
using the option na.action==na.exclude. The summary() function takes as
input an object of class \lm" and provides us with the details of the model-
fitting results, including the overall F-test for association.

```{r}
AnovaMod <- lm(Trait~Geno, na.action=na.exclude)
summary(AnovaMod)
```

# K-W test
A Kruskal-Wallis (K-W) test can also be applied and is more appropriate
in small-sample settings in which the assumption of normality may not be
reasonable.

```{r}
kruskal.test(Trait, Geno, na.action=na.exclude)
```

In this example, the K-W test yields a p-value of 0:085, and a similar conclusion
is reached that there is not sufficient evidence of an association between
resistin c180g and percentage change in muscle strength of the nondominant
arm as measured by NDRM.CH.


## regression


Alternatively, we can fit a multivariable model to either quantitative
or binary traits. Multivariable models have the primary advantage of
allowing us to account appropriately for multiple potential confounders and
effect modifiers.

# Simple and multivariable linear regression
g(E[y]) = XB

\[
g(E(y)) = X\cdot \beta
\]

\[
\mathbf y = \mathbf X \mathbf \beta + \mathbf \epsilon
\]

\[
y_i = \beta_0 + \beta_1 x_i + \epsilon_i
\]


For example, suppose we have m covariates, given by
zi1; : : : ; zim for the ith individual. For example, zi1 may be gender and zi2
may represent smoking status for individual i. We are often interested in
fitting a model of the form

\[
\mathbf y = \mathbf \beta_0 + \mathbf \beta_1\mathbf x_i + \sum_{j=1}^{m}{a_jz_{ij}} + \mathbf \beta + \mathbf \epsilon
\]

# linear regression

```{r}

## interaction considered
Geno <- actn3_r577x
Trait <- NDRM.CH
ModFull <- lm(Trait~Geno+Gender+Geno*Gender, na.action=na.exclude)

summary(ModFull)
```

```{r}
## interaction not considered
ModReduced <- lm(Trait~Geno+Gender, na.action=na.exclude)
summary(ModReduced)
```

## using anova to compare the full model and the reduced model
The anova() function in R provides us with an F-test comparing the full and
reduced models. Formally, this is a test of the null hypothesis that there is no
interaction between genotype and gender:

```{r}
anova(ModReduced, ModFull)
```


The F-test statistic with 2 and 597 degrees of freedom is 3:07, with a corresponding
p-value of 0:047. We therefore conclude that there does appear
to be an interaction between gender and genotype based on these data and
under the modeling assumptions.

## prediction

Finally, the predicted change in muscle strength for the non-dominant arm
and corresponding prediction intervals based on the full model and for each
level of genotype and gender can be generated as follows:
```{r}
# new data
NewDat <- data.frame(Geno=rep(c("CC","CT","TT"),2),Gender=c(rep("Female",3), rep("Male",3)))

# prediction
predict.lm(ModFull, NewDat, interval="prediction", level=0.95)
```


## Logistic regression
for binary traits
As described above, the generalized linear model of Equation (2.9) can also
be applied to a binary trait. In this setting, g() is commonly defined as the
logit function, reducing Equation (2.9) to the logistic regression model

## LD

(Measuring LD using D'). In this example, we use the R package
genetics. Loading this package requires several additional R packages
that are not included in the typical R installation, including MASS, combinat,
gdata, gtools and mvtnorm. Additional details on installing new packages
can be found in the appendix. To load the necessary R packages, we use the
library() function as follows:

```{r}
library(genetics)

# create objects of class genotype for each of our SNP variables.
actn3_r577x[1:10]
Actn3Snp1 <- genotype(actn3_r577x,sep="")
Actn3Snp2 <- genotype(actn3_rs540874,sep="")
Actn3Snp1[1:10]

# class of the snp object
class(Actn3Snp1)
LD(Actn3Snp1,Actn3Snp2)$"D'"
```





