---
title: "statistical_genetics_1"
output: html_document
---

## library
```{r}
library(dplyr)

```

## loading the data
```{r}
fms <- read.delim(file="/Users/jerome/Documents/ebook/FMS_data.txt", header=T, sep="\t")\
head(fms)
```

## Data manipulation
```{r}
attach(fms)
```


## Summary
### identify the minor allel and its frequency
```{r}
# The table() function in R outputs the counts of each level of the ordinal
# variable given as its argument.
GenoCount <- summary(actn3_rs540874)
GenoCount

# total number
NumbObs <- sum(!is.na(actn3_rs540874))

# geno freq
GenoFreq <- as.vector(GenoCount/NumbObs)
GenoFreq
```

```{r}
FreqA <- (2*GenoFreq[1] + GenoFreq[2])/2
FreqA
```

```{r}
FreqG <- (GenoFreq[2] + 2*GenoFreq[3])/2
FreqG
```

```{r}
# Thus, we report A is the minor allele at this SNP locus, with a frequency
# of 0.43.
```

```{r}
#install.packages("genetics")
library(genetics)
Geno <- genotype(actn3_rs540874,sep="")
summary(Geno)
```

## test for association
```{r}
NamesEsr1Snps <- names(fms)[substr(names(fms),1,4)=="esr1"]
NamesEsr1Snps

# The genotype matrix can now be defined by selecting the columns of fms that
# correspond to the esr1 SNP names
fmsEsr1 <- fms[,is.element(names(fms),NamesEsr1Snps)]
```

## trait
```{r}
# We define our trait to be an indicator for whether BMI is > 25 at baseline:
Trait <- as.numeric(pre.BMI>25)

```

## contigency table
```{r}
ObsTab <- table(Trait,Geno)
```

## chisq
```{r}
# test function
chisq_test <- function(Geno) {
    ObsTab <- table(Trait,Geno)
    return(chisq.test(ObsTab)$p.value)
}

table(fmsEsr1$esr1_rs1801132)
apply(fmsEsr1,2,chisq_test)
```

## fisher exact test
```{r}
# fisher test
fisher_test <- function(Geno){
    ObsTab <- table(Trait,Geno)
    return(fisher.test(ObsTab)$p.value)
}

apply(fmsEsr1,2,fisher_test)
```


## quantitative trait
```{r}
NamesResistinSnps <- names(fms)[substr(names(fms),1,8)=="resistin"]
fmsResistin <- fms[,is.element(names(fms),NamesResistinSnps)]
```
create a new function that takes a single genotype vector, converts
it to binary elements, and generates p-values based on the t-test of equality
of the mean trait across the resulting two levels. We first call the genetics
package since we are using the allele.names() function within this package

```{r}
library(genetics)
TtestPval <- function(Geno){
    alleleMajor <- allele.names(genotype(Geno, sep="",
    reorder="freq"))[1]
    GenoWt <- paste(alleleMajor, alleleMajor, sep="")
    GenoBin <- as.numeric(Geno!=GenoWt)[!is.na(Geno)]
    Trait <- NDRM.CH[!is.na(Geno)]
    return(t.test(Trait[GenoBin==1], Trait[GenoBin==0])$p.value)
}

# genoBIn binary vector indicating whether the geno type is wild type or not
# run the test
apply(fmsResistin,2,TtestPval)

```




